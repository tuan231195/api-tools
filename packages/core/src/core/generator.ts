import { ParsedOperationSchema, Schema } from 'src/types';
import dtsGenerator, { Options } from 'dtsgenerator';

export const generate = async (
	parsedSchema: ParsedOperationSchema | Schema[],
	config: Options['config'] = {}
) => {
	const contents = Array.isArray(parsedSchema)
		? parsedSchema
		: ([
				parsedSchema.headers,
				parsedSchema.queries,
				parsedSchema.params,
				parsedSchema.requestBody?.schema,
				...Object.values(parsedSchema.responseBody.statuses),
		  ].filter(Boolean) as Schema[]);
	return dtsGenerator({ contents, config });
};
