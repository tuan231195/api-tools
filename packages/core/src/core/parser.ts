import SwaggerParser from '@apidevtools/swagger-parser';
import { parseSchema } from 'dtsgenerator';
import { omit, omitBy, startCase } from 'lodash';
import { OpenAPI } from 'openapi-types';
import { HTTP_METHODS, JSON_SCHEMA_7 } from 'src/constants';
import logger from 'src/logger';
import {
	JsonSchema,
	Parameter,
	ParsedSchema,
	ParsedOperationSchema,
	ParsedRequestBodySchema,
	ParsedResponseBodySchema,
	Schema,
	SchemaObject,
	supportedMediaTypes,
	ParsedDocument,
} from 'src/types';

export async function validate(uri: string): Promise<OpenAPI.Document> {
	return SwaggerParser.validate(uri);
}

export const parse = async (
	uri: string | OpenAPI.Document
): Promise<ParsedDocument> => {
	const document = typeof uri === 'string' ? await validate(uri) : uri;
	const result: ParsedSchema = {};

	await Promise.all(
		Object.keys(document.paths).map(async (pathName) => {
			const schemaPathComponent = document.paths[pathName] || {};
			for (const methodName of HTTP_METHODS) {
				const operationSchemaComponent =
					schemaPathComponent[methodName];
				if (!operationSchemaComponent) {
					continue;
				}
				if (!operationSchemaComponent.operationId) {
					logger(
						`Skip ${methodName.toUpperCase()} ${pathName}. Operation does not have operationId`
					);
					continue;
				}
				const defaultParameters = schemaPathComponent['parameters'];
				result[operationSchemaComponent.operationId] = {
					schema: await generateApiPathSchema(
						operationSchemaComponent,
						defaultParameters
					),
					info: {
						operationId: operationSchemaComponent.operationId,
						operation: operationSchemaComponent,
						path: pathName,
						method: methodName,
					},
				};
			}
		})
	);

	return {
		document,
		schema: result,
	};
};

const generateApiPathSchema = async (
	schema: OpenAPI.Operation,
	defaultParameters: OpenAPI.Parameters = []
): Promise<ParsedOperationSchema> => {
	logger(`Processing ${schema.operationId}`);

	const operationParameters = schema.parameters ?? defaultParameters;
	const parametersSchema = generateParametersSchema(
		operationParameters as Parameter[]
	);
	const requestBodySchema = generateRequestBodySchema(schema);
	const responseBodySchema = generateResponseBodySchema(schema);
	return {
		headers: parametersSchema['header'] ?? null,
		queries: parametersSchema['query'] ?? null,
		params: parametersSchema['path'] ?? null,
		requestBody: requestBodySchema,
		responseBody: responseBodySchema,
	};
};

const generateRequestBodySchema = (
	schema: OpenAPI.Operation
): ParsedRequestBodySchema | null => {
	let requestBody = getRequestBodySchema(schema);
	if (requestBody) {
		return requestBody;
	}
	requestBody = getBodyParameter(schema);
	return requestBody;
};

const generateResponseBodySchema = (
	operationSchema: OpenAPI.Operation
): ParsedResponseBodySchema => {
	if (!operationSchema.responses) {
		return {
			statuses: {},
			all: null,
			success: null,
		};
	}
	const allStatusSchema = getAllStatusSchema(operationSchema);
	const allSchema: Schema = toSchema({
		$id: 'Response',
		oneOf: Object.values(allStatusSchema).map((response) => ({
			$ref: response.content.$id,
		})),
	});

	const successResponses: Schema[] = Object.entries(allStatusSchema)
		.filter(([status]) => {
			const statusCode = parseInt(status, 10);
			return statusCode >= 200 && statusCode < 300;
		})
		.map(([, schema]) => schema);

	let successSchema: Schema | null = null;
	if (successResponses.length > 0) {
		successSchema = toSchema({
			$id: 'ResponseOk',
			oneOf: successResponses.map((response) => {
				return {
					$ref: response.content.$id,
				};
			}),
		});
	}

	return {
		all: allSchema,
		success: successSchema,
		statuses: allStatusSchema,
	};
};

export const getAllStatusSchema = (
	operationSchema: OpenAPI.Operation
): ParsedResponseBodySchema['statuses'] => {
	return Object.entries(operationSchema.responses || []).reduce(
		(agg, [status, response]) => {
			const statusCode =
				status === 'default' ? 200 : parseInt(status, 10);
			const id = `Response${statusCode}`;
			let jsonBody: any;
			if (response.content && response.content['application/json']) {
				jsonBody = response.content['application/json'];
			} else if (response.schema) {
				jsonBody = response;
			}

			const schema = jsonBody
				? toSchema({
						...jsonBody.schema,
						$id: id,
						example: jsonBody.example ?? null,
				  })
				: nullSchema(id);
			return {
				...agg,
				[statusCode]: schema,
			};
		},
		{} as ParsedResponseBodySchema['statuses']
	);
};

const getRequestBodySchema = (
	operationSchema: OpenAPI.Operation
): ParsedRequestBodySchema | null => {
	const content = (operationSchema as any).requestBody?.content;
	const requestContentType = supportedMediaTypes.find(
		(supportedType) => content && content[supportedType]
	);
	const requestBody = content && content[requestContentType || ''];
	if (!(requestBody && requestContentType)) {
		return null;
	}
	return {
		type: requestContentType === 'application/json' ? 'json' : 'form',
		contentType: requestContentType,
		schema: toSchema({
			...requestBody.schema,
			$id: 'RequestBody',
			example: requestBody.example ?? null,
		}),
	};
};

const getBodyParameter = (
	operationSchema: OpenAPI.Operation
): ParsedRequestBodySchema | null => {
	const parameterGroup = getParameterGroup(
		(operationSchema.parameters || []) as Parameter[]
	);

	let schema: Schema | null = null;
	let requestType: ParsedRequestBodySchema['type'] | null = null;
	let requestContentType: ParsedRequestBodySchema['contentType'] | null =
		null;
	if (parameterGroup['body'] && parameterGroup['body'].length > 0) {
		requestType = 'json';
		requestContentType = 'application/json';
		const bodyParameter = parameterGroup['body'][0];
		schema = toSchema({
			...bodyParameter.schema,
			$id: 'RequestBody',
			example: bodyParameter.example ?? null,
		});
	} else if (parameterGroup['formData']) {
		requestType = 'form';
		requestContentType =
			'consumes' in operationSchema &&
			(operationSchema.consumes || []).includes('multipart/form-data')
				? 'multipart/form-data'
				: 'application/x-www-form-urlencoded';
		const bodyParameters = parameterGroup['formData'];
		schema = parseParameters('RequestBody', bodyParameters);
	}

	if (!(schema && requestType && requestContentType)) {
		return null;
	}

	return {
		type: requestType,
		contentType: requestContentType,
		schema,
	};
};

const getParameterGroup = (parameters: Parameter[]) => {
	const parameterGroup: Record<string, Parameter[]> = {};
	for (const parameter of parameters) {
		const schemaType = parameter.in;
		(parameterGroup[schemaType] = parameterGroup[schemaType] || []).push(
			parameter
		);
	}
	return parameterGroup;
};

const generateParametersSchema = (parameters: Parameter[]) => {
	const parameterGroup = getParameterGroup(parameters);
	// we don't want body parameters
	delete parameterGroup['body'];
	delete parameterGroup['formData'];

	return Object.entries(parameterGroup).reduce((agg, [group, parameters]) => {
		const id = startCase(group);
		return {
			...agg,
			[group]: parseParameters(id, parameters),
		};
	}, {} as Record<string, Schema>);
};

const parseParameters = (id: string, parameters: Parameter[]) => {
	const parameterProperties = parameters.reduce((agg, parameter) => {
		const parameterSchema =
			'schema' in parameter
				? parameter.schema
				: omit(parameter, ['collectionFormat', 'in', 'name']);

		return {
			...agg,
			[parameter.name]: parameterSchema,
		};
	}, {} as Record<string, SchemaObject>);

	const requiredProps = parameters
		.filter((parameter) => parameter.required)
		.map((parameter) => parameter.name);

	return toSchema({
		type: 'object',
		$id: id,
		properties: parameterProperties as any,
		additionalProperties: false,
		required: requiredProps,
	});
};

const nullSchema = (id: string) => {
	return toSchema({
		$id: id,
		type: 'boolean',
		not: true,
	});
};

const toSchema = (schema: JsonSchema): Schema => {
	return parseSchema({
		...omitBy(schema, (value) => value == null),
		$schema: JSON_SCHEMA_7,
	}) as Schema;
};
