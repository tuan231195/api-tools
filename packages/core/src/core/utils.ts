import { zipObject } from 'lodash';

type Awaited<T> = T extends PromiseLike<infer U> ? Awaited<U> : T;

type ResolvedPromise<T extends Record<string, any>> = {
	[k in keyof T]: Awaited<T[k]>;
};

export const promiseAll = async <T extends Record<string, any>>(
	object: T
): Promise<ResolvedPromise<T>> => {
	const keys = Object.keys(object);
	const resolvedArray = await Promise.all(keys.map((key) => object[key]));
	return zipObject(keys, resolvedArray) as ResolvedPromise<T>;
};
