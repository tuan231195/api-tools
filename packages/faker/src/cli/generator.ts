import { CommandModule } from 'yargs';
import { writeFake } from 'src/core';

const commandModule: CommandModule<
	{},
	{
		outDir: string;
		uri: string;
	}
> = {
	describe: 'generate fake data for api',
	handler: async function ({ uri, outDir }) {
		return writeFake(uri, outDir);
	},
	command: '$0',
	builder: {
		outDir: {
			demandOption: true,
			type: 'string',
			description: 'The output directory',
			alias: 'o',
		},
		uri: {
			demandOption: true,
			type: 'string',
			description: 'The swagger file uri',
			alias: 'u',
		},
	},
};

export default commandModule;
