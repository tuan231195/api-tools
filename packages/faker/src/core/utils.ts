import fs from 'fs-extra';
import prettier from 'prettier';
import { transform, isObject } from 'lodash';

export const prepare = async (outDir: string) => {
	await fs.mkdirp(outDir);
};

export const prettify = (content: string) => {
	return prettier.format(content, {
		semi: true,
		parser: 'typescript',
		useTabs: true,
		tabWidth: 4,
		singleQuote: true,
	});
};

export const replaceKeys = (object: any, keyMap: Record<string, string>) => {
	if (!object) {
		return object;
	}
	return transform(object, (result: any, value: any, key: string) => {
		const currentKey = keyMap[key] || key;
		result[currentKey] = isObject(value)
			? replaceKeys(value, keyMap)
			: value;
	});
};
