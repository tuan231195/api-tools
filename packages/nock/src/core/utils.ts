export function wrap<W extends object, B>(wrapper: W, base: B): W & B {
	const proxy = new Proxy(wrapper, {
		get(target, property) {
			if (property in target) {
				return getProperty(target, property, proxy);
			}

			return getProperty(base, property, proxy);
		},
	}) as W & B;
	return proxy;
}

export function getProperty(
	object: any,
	property: string | symbol,
	proxy: any
) {
	const propertyValue: any = object[property];
	if (typeof propertyValue === 'function' && !propertyValue.__proxy) {
		const oldFunction = propertyValue;
		object[property] = function (...args: any[]) {
			const value = oldFunction.apply(object, args);
			return value === object ? proxy : value;
		} as any;

		propertyValue.__proxy = true;
	}
	return propertyValue;
}
