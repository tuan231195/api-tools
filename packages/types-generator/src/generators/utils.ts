import prettier from 'prettier';
import fs from 'fs-extra';
import {
	generate,
	ParsedSchemaOperation,
	Schema,
} from '@vdtn359/api-tools-core';
import os from 'os';

export const prettify = (content: string) => {
	return prettier.format(content, {
		semi: true,
		parser: 'typescript',
		useTabs: true,
		tabWidth: 4,
		singleQuote: true,
	});
};

export const prepare = async (outDir: string) => {
	await fs.mkdirp(outDir);
};

export const toTsType = async (schema: Schema | null) => {
	if (!schema) {
		return null;
	}
	return toTsTypes([schema]);
};

export const toTsTypes = async (schema: ParsedSchemaOperation | Schema[]) => {
	return (await generate(schema))
		.replace(/^declare interface/gm, 'export interface')
		.replace(/^declare type/gm, 'export type')
		.replace(/(\r?\nexport)/g, (match) => `${os.EOL}${match}`);
};
