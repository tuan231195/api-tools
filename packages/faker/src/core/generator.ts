import { kebabCase, omitBy } from 'lodash';
import {
	parse,
	ParsedOperationSchema,
	promiseAll,
} from '@vdtn359/api-tools-core';
import path from 'path';
import { prepare, prettify } from 'src/core/utils';
import logger from 'src/logger';
import fs from 'fs-extra';
import os from 'os';
import { generate } from 'src/core/jsf';
import { FakeOperation } from 'src/core/types';

export const fakeOperation = async (
	operationSchema: ParsedOperationSchema
): Promise<FakeOperation> => {
	const defaultResponse = Object.entries(
		operationSchema.responseBody.statuses
	).find(([key]) => {
		const statusCode = parseInt(key, 10);
		return statusCode >= 200 && statusCode < 300;
	});
	return promiseAll({
		headers: operationSchema.headers
			? generate(operationSchema.headers.content)
			: null,
		queries: operationSchema.queries
			? generate(operationSchema.queries.content)
			: null,
		params: operationSchema.params
			? generate(operationSchema.params.content)
			: null,
		requestBody: operationSchema.requestBody
			? generate(operationSchema.requestBody.schema.content)
			: null,
		responseBody: defaultResponse
			? generate(defaultResponse[1].content)
			: null,
	});
};

export const writeFake = async (uri: string, outDir: string) => {
	const parsedSchema = await parse(uri);
	await prepare(outDir);

	for (const [operationId, { schema: operationSchema }] of Object.entries(
		parsedSchema
	)) {
		try {
			const writeFile = path.resolve(
				outDir,
				`${kebabCase(operationId)}.ts`
			);
			logger(`Write ${operationId} to file ${writeFile}`);

			const exampleObject = omitBy(
				await fakeOperation(operationSchema),
				(value) => value == null
			);

			const output = Object.entries(exampleObject)
				.map(([type, example]) => {
					return `export const ${type} = ${JSON.stringify(example)}`;
				})
				.join(os.EOL.repeat(2));

			await fs.writeFile(writeFile, prettify(output));
		} catch (e) {
			console.error(`Failed to generate fake data for ${operationId}`);
		}
	}
};
