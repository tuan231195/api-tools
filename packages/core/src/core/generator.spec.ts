import path from 'path';
import { mapValues } from 'lodash';
import { parse } from './parser';
import { generate } from './generator';
import { promiseAll } from './utils';

const testDir = path.resolve(__dirname, '..', '..', '..', '..', 'test');

describe('generate', () => {
	it.each([
		['/v2/petstore.json', path.resolve(testDir, 'v2', 'petstore.json')],
		['/v3/petstore.json', path.resolve(testDir, 'v3', 'petstore.json')],
	])('%s', async (testName, filePath) => {
		const result = await parse(filePath);
		const mappedResult = await promiseAll(
			mapValues(result.schema, ({ schema }) => generate(schema))
		);

		expect(mappedResult).toMatchSnapshot();
	});
});
