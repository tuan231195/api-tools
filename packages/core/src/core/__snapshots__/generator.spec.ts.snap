// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generate /v2/petstore.json 1`] = `
Object {
  "addPet": "declare interface RequestBody {
    id?: string; // uuid
    /**
     * example:
     * doggie
     */
    name: string;
    category?: {
        id?: string; // uuid
        name?: string;
    };
    photoUrls: string /* uri */[];
    tags?: {
        id?: string; // uuid
        name?: string;
    }[];
    /**
     * pet status in the store
     */
    status?: \\"available\\" | \\"pending\\" | \\"sold\\";
}
declare type Response405 = void;
",
  "createUser": "declare interface RequestBody {
    id: string; // uuid
    /**
     * example:
     * theUser
     */
    username: string;
    /**
     * example:
     * John
     */
    firstName: string;
    /**
     * example:
     * James
     */
    lastName: string;
    /**
     * example:
     * john@email.com
     */
    email?: string;
    /**
     * example:
     * 12345
     */
    password: string;
    /**
     * example:
     * 12345
     */
    phone?: string;
    /**
     * User Status
     */
    userStatus?: boolean;
}
declare type Response200 = void;
",
  "createUsersWithArrayInput": "declare type RequestBody = {
    id: string; // uuid
    /**
     * example:
     * theUser
     */
    username: string;
    /**
     * example:
     * John
     */
    firstName: string;
    /**
     * example:
     * James
     */
    lastName: string;
    /**
     * example:
     * john@email.com
     */
    email?: string;
    /**
     * example:
     * 12345
     */
    password: string;
    /**
     * example:
     * 12345
     */
    phone?: string;
    /**
     * User Status
     */
    userStatus?: boolean;
}[];
declare type Response200 = void;
",
  "createUsersWithListInput": "declare type RequestBody = {
    id: string; // uuid
    /**
     * example:
     * theUser
     */
    username: string;
    /**
     * example:
     * John
     */
    firstName: string;
    /**
     * example:
     * James
     */
    lastName: string;
    /**
     * example:
     * john@email.com
     */
    email?: string;
    /**
     * example:
     * 12345
     */
    password: string;
    /**
     * example:
     * 12345
     */
    phone?: string;
    /**
     * User Status
     */
    userStatus?: boolean;
}[];
declare type Response200 = void;
",
  "deleteOrder": "declare interface Path {
    /**
     * ID of the order that needs to be deleted
     */
    orderId: string; // uuid
}
declare type Response400 = void;
declare type Response404 = void;
",
  "deletePet": "declare interface Header {
    api_key?: string;
}
declare interface Path {
    /**
     * Pet id to delete
     */
    petId: string; // uuid
}
declare type Response400 = void;
declare type Response404 = void;
",
  "deleteUser": "declare interface Path {
    /**
     * The name that needs to be deleted
     */
    username: string;
}
declare type Response400 = void;
declare type Response404 = void;
",
  "findPetsByStatus": "declare interface Query {
    /**
     * Status values that need to be considered for filter
     */
    status: (\\"available\\" | \\"pending\\" | \\"sold\\")[];
}
declare type Response200 = {
    id?: string; // uuid
    /**
     * example:
     * doggie
     */
    name: string;
    category?: {
        id?: string; // uuid
        name?: string;
    };
    photoUrls: string /* uri */[];
    tags?: {
        id?: string; // uuid
        name?: string;
    }[];
    /**
     * pet status in the store
     */
    status?: \\"available\\" | \\"pending\\" | \\"sold\\";
}[];
declare type Response400 = void;
",
  "findPetsByTags": "declare interface Query {
    /**
     * Tags to filter by
     */
    tags: string[];
}
declare type Response200 = {
    id?: string; // uuid
    /**
     * example:
     * doggie
     */
    name: string;
    category?: {
        id?: string; // uuid
        name?: string;
    };
    photoUrls: string /* uri */[];
    tags?: {
        id?: string; // uuid
        name?: string;
    }[];
    /**
     * pet status in the store
     */
    status?: \\"available\\" | \\"pending\\" | \\"sold\\";
}[];
declare type Response400 = void;
",
  "getInventory": "declare interface Response200 {
    [name: string]: number; // int32
}
",
  "getOrderById": "declare interface Path {
    /**
     * ID of pet that needs to be fetched
     */
    orderId: string; // uuid
}
declare interface Response200 {
    id?: string; // uuid
    petId?: string; // uuid
    quantity?: number; // int32
    shipDate?: string; // date-time
    /**
     * Order Status
     * example:
     * approved
     */
    status?: \\"placed\\" | \\"approved\\" | \\"delivered\\";
    complete?: boolean;
}
declare type Response400 = void;
declare type Response404 = void;
",
  "getPetById": "declare interface Path {
    /**
     * ID of pet to return
     */
    petId: string; // uuid
}
declare interface Response200 {
    id?: string; // uuid
    /**
     * example:
     * doggie
     */
    name: string;
    category?: {
        id?: string; // uuid
        name?: string;
    };
    photoUrls: string /* uri */[];
    tags?: {
        id?: string; // uuid
        name?: string;
    }[];
    /**
     * pet status in the store
     */
    status?: \\"available\\" | \\"pending\\" | \\"sold\\";
}
declare type Response400 = void;
declare type Response404 = void;
",
  "getUserByName": "declare interface Path {
    /**
     * The name that needs to be fetched. Use user1 for testing.
     */
    username: string;
}
declare interface Response200 {
    id: string; // uuid
    /**
     * example:
     * theUser
     */
    username: string;
    /**
     * example:
     * John
     */
    firstName: string;
    /**
     * example:
     * James
     */
    lastName: string;
    /**
     * example:
     * john@email.com
     */
    email?: string;
    /**
     * example:
     * 12345
     */
    password: string;
    /**
     * example:
     * 12345
     */
    phone?: string;
    /**
     * User Status
     */
    userStatus?: boolean;
}
declare type Response400 = void;
declare type Response404 = void;
",
  "loginUser": "declare interface Query {
    /**
     * The user name for login
     */
    username: string;
    /**
     * The password for login in clear text
     */
    password: string;
}
declare type Response200 = string;
declare type Response400 = void;
",
  "logoutUser": "declare type Response200 = void;
",
  "placeOrder": "declare interface RequestBody {
    id?: string; // uuid
    petId?: string; // uuid
    quantity?: number; // int32
    shipDate?: string; // date-time
    /**
     * Order Status
     * example:
     * approved
     */
    status?: \\"placed\\" | \\"approved\\" | \\"delivered\\";
    complete?: boolean;
}
declare interface Response200 {
    id?: string; // uuid
    petId?: string; // uuid
    quantity?: number; // int32
    shipDate?: string; // date-time
    /**
     * Order Status
     * example:
     * approved
     */
    status?: \\"placed\\" | \\"approved\\" | \\"delivered\\";
    complete?: boolean;
}
declare type Response400 = void;
",
  "updatePet": "declare interface RequestBody {
    id?: string; // uuid
    /**
     * example:
     * doggie
     */
    name: string;
    category?: {
        id?: string; // uuid
        name?: string;
    };
    photoUrls: string /* uri */[];
    tags?: {
        id?: string; // uuid
        name?: string;
    }[];
    /**
     * pet status in the store
     */
    status?: \\"available\\" | \\"pending\\" | \\"sold\\";
}
declare type Response400 = void;
declare type Response404 = void;
declare type Response405 = void;
",
  "updatePetWithForm": "declare interface Path {
    /**
     * ID of pet that needs to be updated
     */
    petId: string; // uuid
}
declare interface RequestBody {
    /**
     * Updated name of the pet
     */
    name?: string;
    /**
     * Updated status of the pet
     */
    status?: string;
}
declare type Response405 = void;
",
  "updateUser": "declare interface Path {
    /**
     * name that need to be updated
     */
    username: string;
}
declare interface RequestBody {
    id: string; // uuid
    /**
     * example:
     * theUser
     */
    username: string;
    /**
     * example:
     * John
     */
    firstName: string;
    /**
     * example:
     * James
     */
    lastName: string;
    /**
     * example:
     * john@email.com
     */
    email?: string;
    /**
     * example:
     * 12345
     */
    password: string;
    /**
     * example:
     * 12345
     */
    phone?: string;
    /**
     * User Status
     */
    userStatus?: boolean;
}
declare type Response400 = void;
declare type Response404 = void;
",
  "uploadFile": "declare interface Path {
    /**
     * ID of pet to update
     */
    petId: string; // uuid
}
declare interface RequestBody {
    /**
     * Additional data to pass to server
     */
    additionalMetadata?: string;
    /**
     * file to upload
     */
    file?: unknown;
}
declare interface Response200 {
    code?: number; // int32
    type?: string;
    message?: string;
}
",
}
`;

exports[`generate /v3/petstore.json 1`] = `
Object {
  "addPet": "declare interface RequestBody {
    /**
     * example:
     * 10
     */
    id?: string; // uuid
    /**
     * example:
     * doggie
     */
    name: string;
    category?: {
        id?: string; // uuid
        name?: string;
    };
    photoUrls: string /* uri */[];
    tags?: {
        id?: string; // uuid
        name?: string;
    }[];
    /**
     * pet status in the store
     */
    status?: \\"available\\" | \\"pending\\" | \\"sold\\";
}
declare interface Response200 {
    /**
     * example:
     * 10
     */
    id?: string; // uuid
    /**
     * example:
     * doggie
     */
    name: string;
    category?: {
        id?: string; // uuid
        name?: string;
    };
    photoUrls: string /* uri */[];
    tags?: {
        id?: string; // uuid
        name?: string;
    }[];
    /**
     * pet status in the store
     */
    status?: \\"available\\" | \\"pending\\" | \\"sold\\";
}
declare type Response405 = void;
",
  "createUser": "declare interface RequestBody {
    id: string; // uuid
    /**
     * example:
     * theUser
     */
    username: string;
    /**
     * example:
     * John
     */
    firstName: string;
    /**
     * example:
     * James
     */
    lastName: string;
    /**
     * example:
     * john@email.com
     */
    email?: string;
    /**
     * example:
     * 12345
     */
    password: string;
    /**
     * example:
     * 12345
     */
    phone?: string;
    /**
     * User Status
     */
    userStatus?: boolean;
}
declare interface Response200 {
    id: string; // uuid
    /**
     * example:
     * theUser
     */
    username: string;
    /**
     * example:
     * John
     */
    firstName: string;
    /**
     * example:
     * James
     */
    lastName: string;
    /**
     * example:
     * john@email.com
     */
    email?: string;
    /**
     * example:
     * 12345
     */
    password: string;
    /**
     * example:
     * 12345
     */
    phone?: string;
    /**
     * User Status
     */
    userStatus?: boolean;
}
",
  "createUsersWithListInput": "declare type RequestBody = {
    id: string; // uuid
    /**
     * example:
     * theUser
     */
    username: string;
    /**
     * example:
     * John
     */
    firstName: string;
    /**
     * example:
     * James
     */
    lastName: string;
    /**
     * example:
     * john@email.com
     */
    email?: string;
    /**
     * example:
     * 12345
     */
    password: string;
    /**
     * example:
     * 12345
     */
    phone?: string;
    /**
     * User Status
     */
    userStatus?: boolean;
}[];
declare type Response200 = void;
",
  "deleteOrder": "declare interface Path {
    orderId: string; // uuid
}
declare type Response400 = void;
declare type Response404 = void;
",
  "deletePet": "declare interface Header {
    api_key?: string;
}
declare interface Path {
    petId: string; // uuid
}
declare type Response400 = void;
",
  "deleteUser": "declare interface Path {
    username: string;
}
declare type Response400 = void;
declare type Response404 = void;
",
  "findPetsByStatus": "declare interface Query {
    status?: \\"available\\" | \\"pending\\" | \\"sold\\";
}
declare type Response200 = {
    /**
     * example:
     * 10
     */
    id?: string; // uuid
    /**
     * example:
     * doggie
     */
    name: string;
    category?: {
        id?: string; // uuid
        name?: string;
    };
    photoUrls: string /* uri */[];
    tags?: {
        id?: string; // uuid
        name?: string;
    }[];
    /**
     * pet status in the store
     */
    status?: \\"available\\" | \\"pending\\" | \\"sold\\";
}[];
declare type Response400 = void;
",
  "findPetsByTags": "declare interface Query {
    tags?: string[];
}
declare type Response200 = {
    /**
     * example:
     * 10
     */
    id?: string; // uuid
    /**
     * example:
     * doggie
     */
    name: string;
    category?: {
        id?: string; // uuid
        name?: string;
    };
    photoUrls: string /* uri */[];
    tags?: {
        id?: string; // uuid
        name?: string;
    }[];
    /**
     * pet status in the store
     */
    status?: \\"available\\" | \\"pending\\" | \\"sold\\";
}[];
declare type Response400 = void;
",
  "getInventory": "declare interface Response200 {
    [name: string]: number; // int32
}
",
  "getOrderById": "declare interface Path {
    orderId: string; // uuid
}
declare interface Response200 {
    id?: string; // uuid
    petId?: string; // uuid
    quantity?: number; // int32
    shipDate?: string; // date-time
    /**
     * Order Status
     * example:
     * approved
     */
    status?: \\"placed\\" | \\"approved\\" | \\"delivered\\";
    complete?: boolean;
}
declare type Response400 = void;
declare type Response404 = void;
",
  "getPetById": "declare interface Path {
    petId: string; // uri
}
declare interface Response200 {
    /**
     * example:
     * 10
     */
    id?: string; // uuid
    /**
     * example:
     * doggie
     */
    name: string;
    category?: {
        id?: string; // uuid
        name?: string;
    };
    photoUrls: string /* uri */[];
    tags?: {
        id?: string; // uuid
        name?: string;
    }[];
    /**
     * pet status in the store
     */
    status?: \\"available\\" | \\"pending\\" | \\"sold\\";
}
declare type Response400 = void;
declare type Response404 = void;
",
  "getUserByName": "declare interface Path {
    username: string;
}
declare interface Response200 {
    id: string; // uuid
    /**
     * example:
     * theUser
     */
    username: string;
    /**
     * example:
     * John
     */
    firstName: string;
    /**
     * example:
     * James
     */
    lastName: string;
    /**
     * example:
     * john@email.com
     */
    email?: string;
    /**
     * example:
     * 12345
     */
    password: string;
    /**
     * example:
     * 12345
     */
    phone?: string;
    /**
     * User Status
     */
    userStatus?: boolean;
}
declare type Response400 = void;
declare type Response404 = void;
",
  "loginUser": "declare interface Query {
    username?: string;
    password?: string;
}
declare type Response200 = string;
declare type Response400 = void;
",
  "logoutUser": "declare type Response200 = void;
",
  "placeOrder": "declare interface RequestBody {
    id?: string; // uuid
    petId?: string; // uuid
    quantity?: number; // int32
    shipDate?: string; // date-time
    /**
     * Order Status
     * example:
     * approved
     */
    status?: \\"placed\\" | \\"approved\\" | \\"delivered\\";
    complete?: boolean;
}
declare interface Response200 {
    id?: string; // uuid
    petId?: string; // uuid
    quantity?: number; // int32
    shipDate?: string; // date-time
    /**
     * Order Status
     * example:
     * approved
     */
    status?: \\"placed\\" | \\"approved\\" | \\"delivered\\";
    complete?: boolean;
}
declare type Response405 = void;
",
  "updatePet": "declare interface RequestBody {
    /**
     * example:
     * 10
     */
    id?: string; // uuid
    /**
     * example:
     * doggie
     */
    name: string;
    category?: {
        id?: string; // uuid
        name?: string;
    };
    photoUrls: string /* uri */[];
    tags?: {
        id?: string; // uuid
        name?: string;
    }[];
    /**
     * pet status in the store
     */
    status?: \\"available\\" | \\"pending\\" | \\"sold\\";
}
declare interface Response200 {
    /**
     * example:
     * 10
     */
    id?: string; // uuid
    /**
     * example:
     * doggie
     */
    name: string;
    category?: {
        id?: string; // uuid
        name?: string;
    };
    photoUrls: string /* uri */[];
    tags?: {
        id?: string; // uuid
        name?: string;
    }[];
    /**
     * pet status in the store
     */
    status?: \\"available\\" | \\"pending\\" | \\"sold\\";
}
declare type Response400 = void;
declare type Response404 = void;
declare type Response405 = void;
",
  "updatePetWithForm": "declare interface Path {
    petId: string; // uuid
}
declare interface Query {
    name?: string;
    status?: string;
}
declare type Response405 = void;
",
  "updateUser": "declare interface Path {
    username: string;
}
declare interface RequestBody {
    id: string; // uuid
    /**
     * example:
     * theUser
     */
    username: string;
    /**
     * example:
     * John
     */
    firstName: string;
    /**
     * example:
     * James
     */
    lastName: string;
    /**
     * example:
     * john@email.com
     */
    email?: string;
    /**
     * example:
     * 12345
     */
    password: string;
    /**
     * example:
     * 12345
     */
    phone?: string;
    /**
     * User Status
     */
    userStatus?: boolean;
}
declare interface Response200 {
    id: string; // uuid
}
",
  "uploadFile": "declare interface Path {
    petId: string; // uuid
}
declare interface Query {
    additionalMetadata?: string;
}
declare interface Response200 {
    code?: number; // int32
    type?: string;
    message?: string;
}
",
}
`;
